[GTK4] Handle hold gesture events

alice-mkh opened the pull request:
#### e9ba79dac78dd3f149ea27d52e0b6cb684990e5a
<pre>
[GTK4] Handle hold gesture events
<a href="https://bugs.webkit.org/show_bug.cgi?id=237247">https://bugs.webkit.org/show_bug.cgi?id=237247</a>

Reviewed by Michael Catanzaro.

In GTK 4.7.x, scroll controllers don&apos;t necessarily work with scroll events,
but can also involve hold gesture events, as well as no event at all.

Handle scroll-begin signal, and add safeguards for null and hold events
everywhere.

* Source/WebKit/Shared/gtk/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(handleScroll):
(webkitWebViewBaseScrollBegin):
(webkitWebViewBaseScroll):
(webkitWebViewBaseScrollEnd):
(webkitWebViewBaseConstructed):
* Source/WebKit/UIProcess/gtk/ViewGestureControllerGtk.cpp:
(WebKit::ViewGestureController::SwipeProgressTracker::handleEvent):

Canonical link: <a href="https://commits.webkit.org/253084@main">https://commits.webkit.org/253084@main</a>
</pre>

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1334,27 +1334,37 @@ static gboolean webkitWebViewBaseScroll(WebKitWebViewBase* webViewBase, double d
-
-        bool isEnd = gdk_scroll_event_is_stop(event) ? true : false;
+        GdkDevice* device = gtk_event_controller_get_current_event_device(eventController);
+        // We only have hold events on touchpads, so assume one if we get no event
```
```suggestion
        // We only have hold events on touchpads, so assume one if we get no event.
```

mcatanzaro commented on Source/WebKit/UIProcess/gtk/ViewGestureControllerGtk.cpp:
```diff
@@ -195,10 +195,11 @@ bool ViewGestureController::SwipeProgressTracker::handleEvent(PlatformGtkScrollD
 
     Seconds time = Seconds::fromMilliseconds(eventTime);
-    if (time != m_prevTime)
+    if (time > m_prevTime) {
```
This change is intended to be part of this MR, right?

alice-mkh replied:
Correct. When the event is null, we get timestamp 0 and don't want it here.

alice-mkh replied:
Marking as resolved, feel free to unresolve if you think it warrants a comment.

alice-mkh committed:
[GTK4] Handle hold gesture events
https://bugs.webkit.org/show_bug.cgi?id=237247

Reviewed by Michael Catanzaro.

In GTK 4.7.x, scroll controllers don't necessarily work with scroll events,
but can also involve hold gesture events, as well as no event at all.

Handle scroll-begin signal, and add safeguards for null and hold events
everywhere.

* Source/WebKit/Shared/gtk/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(handleScroll):
(webkitWebViewBaseScrollBegin):
(webkitWebViewBaseScroll):
(webkitWebViewBaseScrollEnd):
(webkitWebViewBaseConstructed):
* Source/WebKit/UIProcess/gtk/ViewGestureControllerGtk.cpp:
(WebKit::ViewGestureController::SwipeProgressTracker::handleEvent):

Canonical link: https://commits.webkit.org/253084@main
